design/batch_retrieval_design.md

```diff
@@ -2,19 +2,140 @@
 
+*Decision:* We decided to go with #2 and create our own API. The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. 
+
+```
+POST /secrets/batch
+
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+### Request Body
+
+The request body is a list of all variable IDs (paths to secrets in Conjur/DAP) that need to be retrieved. For example: 
+
+```
+{
+  "variables": [
+    {"name": "var/path/conjurSecret"},
+    {"name": "var/path/anotherConjurSecret"},
+    ...
+  ]
+}
+```
+
+#### Headers
+
+```
+Content-Type: application/json
+```
+
+### Response
+
+| Code             | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 200 OK           | All secret values were retrieved successfully                |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations) |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
+| 404 Not Found    | At least one variable does not exist, or at least one variable does not have any secret values |
+
+#### Example response
+
+```
+HTTP/1.1 200 OK
+
+{
+  "sucessSecret1":[
+     {
+       "code": 200
+       "value": "superSecret"
+     }
+  ]
+  "failSecret1":[
+    {
+      "code": 401
+      "value": "Unauthorized"     
+    }
+  ]
+  "failSecret2":[
+    {
+      "code": 404
+      "value": "Not Found"    
+    }
+  ]
+  "sucessSecret1":[
+    {
+      "code": 200
+      "value": "superSuperSecret"    
+    }
+  ]
+}
+```
+
+All secret values will be treated as individual entities. If a single secret fails to be retrieved, this will not impact the integrity of the request.
+
+If no Conjur/DAP secrets were able to be retrieved (no 200 code), the Batch Retreival response fail with a 5xx error code.
```

Comment:
why 5xx?

Reply:
I realize now that this is wrong. The response will still return a 200 OK because the request was handled even if there was a problem in getting the value of the secret (404 / 401)